src/game/ZoneManager.cxx

```diff
@@ -3,9 +3,55 @@
 #include "LOG.hxx"
 #include "../services/GameClock.hxx"
 #include "../services/Randomizer.hxx"
+#include "GameStates.hxx"
 
 ZoneManager::ZoneManager()
 {
+  if (!Engine::instance().map)
+  {
+    LOG(LOG_ERROR) << "ZoneManager() constructor has been called with no valid map object available!";
+  }
+
+  Engine::instance().map->registerCallbackFunction(
+      [this](const MapNode &mapNode)
+      {
+        // If we place a building on zone tile, occupy the node
+        if (mapNode.getTileData(Layer::BUILDINGS) && mapNode.getTileData(Layer::ZONE))
+        {
+          for (auto &zoneArea : m_zoneAreas)
+          {
+            zoneArea.occupyZoneNode(mapNode.getCoordinates());
```

Comment:
`isWithinBoundaries` and `isPartOfZone` are your friends :)
ZoneNode can be part only of one ZoneArea, so no need to search further once it found the correct place.

To much brute force Liz :)

Reply:
sorry, I don't understand what you mean here at all 🙈

Reply:
please check my new implementation